internal/bminventory/inventory.go

```diff
@@ -1553,27 +1553,83 @@ func (b *bareMetalInventory) GetClusterSupportedPlatforms(ctx context.Context, p
+			log.Error("UpdateClusterInstallConfigInternal failed")
+			tx.Rollback()
+		}
+		if r := recover(); r != nil {
+			log.Error("UpdateClusterInstallConfigInternal failed to recover")
+			tx.Rollback()
+		}
+	}()
+
+	// in case host monitor already updated the state we need to use FOR UPDATE option
+	if cluster, err = common.GetClusterFromDBForUpdate(tx, params.ClusterID, common.UseEagerLoading); err != nil {
 		log.WithError(err).Errorf("failed to find cluster %s", params.ClusterID)
 		return nil, err
 	}
 
-	if err = b.installConfigBuilder.ValidateInstallConfigPatch(&cluster, params.InstallConfigParams); err != nil {
+	if err = b.installConfigBuilder.ValidateInstallConfigPatch(cluster, params.InstallConfigParams); err != nil {
 		return nil, common.NewApiError(http.StatusBadRequest, err)
 	}
 
-	err = b.db.Model(&common.Cluster{}).Where(query, params.ClusterID).Update("install_config_overrides", params.InstallConfigParams).Error
+	// Set install config overrides feature usage
+	err = b.setInstallConfigOverridesUsage(cluster.Cluster.FeatureUsage, params.InstallConfigParams, params.ClusterID, tx)
 	if err != nil {
-		return nil, common.NewApiError(http.StatusInternalServerError, err)
+		log.WithError(err).Errorf("failed to set install config overrides feature usage for cluster %s", params.ClusterID)
 	}
 
+	err = tx.Model(&common.Cluster{}).Where(query, params.ClusterID).Update("install_config_overrides", params.InstallConfigParams).Error
+	if err != nil {
+		log.WithError(err).Errorf("failed to update install config overrides")
+		return nil, common.NewApiError(http.StatusInternalServerError, err)
+	}
+	err = tx.Commit().Error
+	if err != nil {
+		log.Error(err)
+		return nil, common.NewApiError(http.StatusInternalServerError, err)
+	}
+	txSuccess = true
 	eventgen.SendInstallConfigAppliedEvent(ctx, b.eventsHandler, params.ClusterID)
-	log.Infof("Custom install config was applied to cluster %s", params.ClusterID)
-	return &cluster, nil
+	return cluster, nil
+}
+
+func (b *bareMetalInventory) setInstallConfigOverridesUsage(featureUsages string, installConfigParams string, clusterID strfmt.UUID, db *gorm.DB) error {
+	usages, err := usage.Unmarshal(featureUsages)
+	if err != nil {
+		return err
+	}
+	var installConfigOverrides map[string]interface{}
+	err = json.Unmarshal([]byte(installConfigParams), &installConfigOverrides)
+	if err != nil {
+		return err
+	} else if len(installConfigOverrides) < 1 {
+		return nil
+	}
+	props := usages[usage.InstallConfigOverrides].Data
+	if props == nil {
+		props = make(map[string]interface{})
+	}
+	for installConfigKey, installConfigValue := range installConfigOverrides {
+		key := installConfigKey
+		switch secondaryKeys := installConfigValue.(type) {
+		case map[string]interface{}:
+			for secondaryKey := range secondaryKeys {
+				key = key + " " + secondaryKey
+			}
+		}
+		props[key] = true
+	}
+	b.setUsage(true, usage.InstallConfigOverrides, &props, usages)
+	b.usageApi.Save(db, clusterID, usages)
```

Comment:
Looks like save is also doing a database UPDATE

Reply:
But @CrystalChun added another update, it's just redundant:
![image](https://user-images.githubusercontent.com/10882062/166699950-c76cc457-e963-4a41-857c-5d105ba9b7fc.png)

Reply:
Oh the update in the screenshot is for a different property [`install_config_overrides`](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/bminventory/inventory.go#L1589) whereas the usage api's save's update is for the [`feature usage`](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/usage/manager.go#L55) property. Should there only be one update in this case for all the changes?

Reply:
The same type of double update happens in `v2UpdateClusterInternal` too:
1. Call to [`UpdateClusterData`](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/bminventory/inventory.go#L1874) which also [updates multiple properties](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/bminventory/inventory.go#L2165) separately
1. Call to [feature usage save](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/bminventory/inventory.go#L1895)

Does this need to be modified as well?

Reply:
>Oh the update in the screenshot is for a different property [install_config_overrides](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/bminventory/inventory.go#L1589) whereas the usage api's save's update is for the [feature usage](https://github.com/openshift/assisted-service/blob/a96f843cb8703f4e8102cc6d3db081a527eca6f6/internal/usage/manager.go#L55) property

Oops yeah completely missed that 

>Should there only be one update in this case for all the changes?

No sorry please ignore me

Reply:
No worries! They do seem quite similar. Thank you for your reviews!